src/hash_witness.rs

```diff
@@ -0,0 +1,325 @@
+        for (slot_name, p) in &self.slots {
+            if slot_name == name {
+                return *p;
+            }
+        }
+
+        Stub::Dummy
+    }
+
+    pub fn car_cdr_named(
+        &mut self,
+        name: ConsName,
+        store: &mut Store<F>,
+        cons: &Ptr<F>,
+    ) -> (Ptr<F>, Ptr<F>) {
+        self.get_assigned_slot(name).car_cdr(store, cons)
+    }
+
+    pub fn cons_named(
+        &mut self,
+        name: ConsName,
+        store: &mut Store<F>,
+        car: Ptr<F>,
+        cdr: Ptr<F>,
+    ) -> Ptr<F> {
+        self.get_assigned_slot(name).cons(store, car, cdr)
+    }
+
+    pub fn strcons_named(
+        &mut self,
+        name: ConsName,
+        store: &mut Store<F>,
+        car: Ptr<F>,
+        cdr: Ptr<F>,
+    ) -> Ptr<F> {
+        self.get_assigned_slot(name).strcons(store, car, cdr)
+    }
+
+    pub fn car_cdr_mut_named(
+        &mut self,
+        name: ConsName,
+        store: &mut Store<F>,
+        cons: &Ptr<F>,
+    ) -> Result<(Ptr<F>, Ptr<F>), String> {
+        self.get_assigned_slot(name).car_cdr_mut(store, cons)
+    }
+
+    pub fn extend_named(
+        &mut self,
+        name: ConsName,
+        env: Ptr<F>,
+        var: Ptr<F>,
+        val: Ptr<F>,
+        store: &mut Store<F>,
+    ) -> Ptr<F> {
+        let binding = self.cons_named(ConsName::Binding, store, var, val);
+
+        self.cons_named(name, store, binding, env)
+    }
+
+    pub fn assert_invariants(&self, _store: &Store<F>) {
+        // Use the commented code below to search for (non-nil) duplicated conses, which could indicate that two
+        // different Cons are being used to reference the identical structural value. In that case, they could be
+        // coalesced, potentially leading to fewer slots being required.
+        //
+        // As of the initial optimization pass, Env and ExtendedClosureEnv appear to be duplicated in this way. However,
+        // it's not clear why that is, and coalescing them does not obviously lead to a potential savings, so we will
+        // leave them distinct for now.
+
+        // let mut digests = HashMap::new();
+
+        // for (name, p) in self.slots.iter() {
+        //     match p {
+        //         Stub::Value(hash) => {
+        //             if let Some(existing_name) = digests.insert(hash.cons, name) {
+        //                 use crate::writer::Write;
+        //                 let nil = store.get_nil();
+        //                 if !store.ptr_eq(&hash.cons, &nil).unwrap() {
+        //                     let cons = hash.cons.fmt_to_string(store);
+        //                     dbg!(hash.cons, cons, name, existing_name);
+        //                     panic!("duplicate");
+        //                 }
+        //             };
+        //         }
+        //         _ => (),
+        //     };
+        // }
+
+        assert!(self.conses_used_count() <= MAX_CONSES_PER_REDUCTION);
```

Comment:
This is somewhat redundant now that there is a fixed set of slots, which make it impossible that more than that number be used. This particular assertion is an artifact of development and precedes the fixed allocation scheme. Nevertheless, the invariant should still remain correct; and this method `assert_invariants` is important at least for debugging and development. We can rework this particular assertion in the future, to avoid the cost.